design/open/commit-metadata/airflow.md

```diff
@@ -0,0 +1,180 @@
+
+### Source of truth
+
+A [wise woman once wrote](https://twitter.com/mipsytipsy/status/998084191488126976):
+
+> You either have one source of truth, or multiple sources of lies.
+
+There is a natural tension and thereby a continuoum between these two
+extremes:
+
+* **Normalized data.**  lakeFS commits hold only enough to identify an
+  Airflow run - its URL.  Airflow holds all other run metadata, and is
+  required to make use of the URL.
+* **Denormalized data.**  lakeFS commits hold all the information that
+  lakeFS and its users need about an Airflow run.  All run metadata is
+  copied to lakeFS commit metadata.
+
+We always need the normalized data pointer, it is the only real identifier
+of the run.[^1] The lakeFS UI can offer the best integration when lakeFS is
+the source of truth.  Initially we shall allow users to select what metadata
+to copy.  The default will be _all_.  For velocity we will avoid adding any
+metadata that is in any way difficult to produce in the lakeFS Airflow
+commit operator.  We may revisit this decision after feedback from users.
+
+[^1]: Naively the run ID would be sufficient.  But that imposes an
+    assumption of a single Airflow instance.
+
+### Scope
+
+We add to each commit metadata for these items:
+
+* DAG run.  The URL of the DAG run in Airflow that generated this commit.
+  This is the URL of [Get a DAG run][get-dag-run] in the REST API.
+* Optionally, additional DAG identifiers such as:
+  * Git commit digest
+  * Airflow system identifier
+  * Airflow logical ("execution") date
+  * Notes (an Airflow metadata concept)
+  * Conf.
+
+  We allow all items that can be retrieved by [Get a DAG run][get-dag-run]
+  in the REST API, and copy terminology from there.
+
+  All optionals must be exactly that.  Most importantly, conf can contain
+  secrets and _must_ be optional.
+  
+### Naming
+
+We define a simple naming scheme that allows some future-compatibility.  All
+items use the prefix "`::lakefs::Airflow::`", staking a claim to the
+"`::lakefs::`" prefix on commit metadata.
+
+lakeFS supports only string metadata.  All plain suffixes are assumed to be
+plain strings.  If a suffix ends in `[...]`, that encodes a _type_.  We
+initially define the types that are needed to integrate Airflow.
+
+We add these suffixes to encode how to access the DAG run:
+
+* `url[url:id]`: The DAG run
+* `url[url:ui]`: URL to access the Airflow UI for the run.
+
+The REST API suggests the following suffixes:
+
+* `dag_run_id`: string
+* `dag_id`: string
+* `logical_date[iso8601]`: ISO 8601
+
+  `execution_date` is deprecated by Airflow and equal to `logical_date`, so
+  it is _not_ included.
+* `data_interval_start[iso8601]`: ISO 8601
+* `data_interval_end[iso8601]`: ISO 8601
+* `last_scheduling_decision[iso8601]`:  ISO 8601
+* `run_type`: string
+* `external_trigger[boolean]`: boolean encoded as "`true`" or "`false`"
+* `conf[json]`: string encoding a JSON object of conf
+* `note`: string
```

Comment:
Can you add an example of a lakeFS commit metadata with all these values, I'm not sure I follow the naming convention and the parsing.

Reply:
Done!